From 5d50c35a4c88b99ae4bf6f934cbe23a80f176174 Mon Sep 17 00:00:00 2001
From: yoinx <joseph.schubert@gmail.com>
Date: Thu, 9 Feb 2017 18:02:19 -0500
Subject: [PATCH 2/2] Add Functionality to control High Brightness Mode [3/3]

This commit will allow the rom to control this kernel based feature without the need for root.

Relies on changes in frameworks/base.

Code adapted from dhacker29's commit here: PureNexusProject-Legacy/android_device_moto_shamu@ecfb47e

Change-Id: I6547b4cb0c6bae150f2aeb514f0f18c6088b7738
Signed-off-by: Nathan Chancellor <natechancellor@gmail.com>
---
 init.angler.rc                                         |  3 +++
 overlay/frameworks/base/core/res/res/values/config.xml |  4 ++++
 power/power.c                                          | 15 +++++++++++++++
 sepolicy/file.te                                       |  2 ++
 sepolicy/file_contexts                                 |  2 ++
 sepolicy/system_server.te                              |  2 ++
 6 files changed, 28 insertions(+)

diff --git a/init.angler.rc b/init.angler.rc
index 23a1822..fd4ce5e 100755
--- a/init.angler.rc
+++ b/init.angler.rc
@@ -137,6 +137,9 @@ on boot
     write /sys/block/mmcblk0/queue/rq_affinity 0
     write /sys/block/mmcblk0/queue/scheduler maple
 
+    # High Brightness Mode
+    chown system system /sys/devices/virtual/graphics/fb0/hbm
+
 on property:init.svc.per_mgr=running
     start per_proxy
 
diff --git a/overlay/frameworks/base/core/res/res/values/config.xml b/overlay/frameworks/base/core/res/res/values/config.xml
index b6303d7..65a120f 100755
--- a/overlay/frameworks/base/core/res/res/values/config.xml
+++ b/overlay/frameworks/base/core/res/res/values/config.xml
@@ -444,4 +444,8 @@
 
     <!-- Whether the device needs axis inversion for the fingerprint notification panel gesture when rotated -->
     <bool name="config_needsFingerprintAxisInversion">true</bool>
+
+    <!-- Whether device supports High Brightness Mode -->
+    <bool name="config_supportHighBrightness">true</bool>
+
 </resources>
diff --git a/power/power.c b/power/power.c
index cc1aba2..c5a667f 100644
--- a/power/power.c
+++ b/power/power.c
@@ -56,6 +56,8 @@
 #define CPU5_ONLINE_PATH "/sys/devices/system/cpu/cpu5/online"
 #define CPU6_ONLINE_PATH "/sys/devices/system/cpu/cpu6/online"
 #define CPU7_ONLINE_PATH "/sys/devices/system/cpu/cpu7/online"
+#define HIGH_BRIGHTNESS_MODE_PATH "/sys/devices/virtual/graphics/fb0/hbm"
+#define POWER_FEATURE_HIGH_BRIGHTNESS_MODE 1 // this must be the same as what is being passed from the rom's powerhal
 
 #define PLATFORM_SLEEP_MODES 2
 #define XO_VOTERS 3
@@ -249,6 +251,18 @@ static long long calc_timespan_us(struct timespec start, struct timespec end) {
     return diff_in_us;
 }
 
+static void set_feature(struct power_module *module, feature_t feature, int state)
+{
+    switch (feature) {
+    case POWER_FEATURE_HIGH_BRIGHTNESS_MODE:
+        sysfs_write(HIGH_BRIGHTNESS_MODE_PATH, state ? "1" : "0");
+        break;
+    default:
+        ALOGW("Error setting the feature, it doesn't exist %d\n", feature);
+        break;
+    }
+}
+
 static void power_hint(struct power_module *module, power_hint_t hint,
         void *data)
 {
@@ -763,6 +777,7 @@ struct power_module HAL_MODULE_INFO_SYM = {
 
     .init = power_init,
     .powerHint = power_hint,
+    .setFeature = set_feature,
     .setInteractive = set_interactive,
     .get_number_of_platform_modes = get_number_of_platform_modes,
     .get_platform_low_power_stats = get_platform_low_power_stats,
diff --git a/sepolicy/file.te b/sepolicy/file.te
index 74484b4..0456637 100644
--- a/sepolicy/file.te
+++ b/sepolicy/file.te
@@ -35,3 +35,5 @@ type cnd_data_file, file_type, data_file_type;
 
 type sysfs_wifi_sar, fs_type, sysfs_type;
 type sysfs_lcd_mipi, fs_type, sysfs_type;
+
+type higbrightness_dev, fs_type, sysfs_type;
diff --git a/sepolicy/file_contexts b/sepolicy/file_contexts
index a8a68a6..e91e24c 100644
--- a/sepolicy/file_contexts
+++ b/sepolicy/file_contexts
@@ -173,3 +173,5 @@
 # Sysfs files used by nanoapp_cmd
 /sys/devices/virtual/nanohub/nanohub(/.*)? u:object_r:sysfs_nanoapp_cmd:s0
 
+# High Brightess Mode
+/sys/devices/virtual/graphics/fb0/hbm               u:object_r:higbrightness_dev:s0
diff --git a/sepolicy/system_server.te b/sepolicy/system_server.te
index ac98a56..4f71292 100644
--- a/sepolicy/system_server.te
+++ b/sepolicy/system_server.te
@@ -44,3 +44,5 @@ allow system_server graphics_device:chr_file rw_file_perms;
 
 # Access for thermal-engine
 allow system_server sysfs_thermal:file write;
+
+allow system_server higbrightness_dev:file rw_file_perms;
-- 
2.13.0

